### PR TITLE
Improve the rewriter for SINE.

### PR DESCRIPTION
**Pull request by ajreynol**
This improves the arith rewriter to ensure factors of PI are between -1 and 1 in arguments of SINE.

This is required for #1168.

### CODE REVIEW

File: src/theory/arith/arith_rewriter.cpp [hunk 1]
```diff
@@ -384,15 +384,40 @@ RewriteResponse ArithRewriter::postRewriteTranscendental(TNode t) {
...
+          if( monomial.getKind()==kind::PI ){
+            pi_factor = NodeManager::currentNM()->mkConst(Rational(1));
+            pi = monomial;
+          }else if( monomial.getKind()==kind::MULT && monomial[0].isConst() && monomial[1].getKind()==kind::PI ){
+            pi_factor = monomial[0];
+            pi = monomial[1];
+          }
+          if( !pi_factor.isNull() ){
+            // remainder is the sum of the child besides this one
+            std::vector< Node > remc;
```
Comment (timothy-king) [inline src/theory/arith/arith_rewriter.cpp]: Move the logic to generate rem from j and t[0] into its own function.

File: src/theory/arith/arith_rewriter.cpp [hunk 2]
```diff
@@ -384,15 +384,40 @@ RewriteResponse ArithRewriter::postRewriteTranscendental(TNode t) {
...
+      Node rem;
       if( t[0].getKind()==kind::PI ){
         pi_factor = NodeManager::currentNM()->mkConst(Rational(1));
         pi = t[0];
       }else if( t[0].getKind()==kind::MULT && t[0][0].isConst() && t[0][1].getKind()==kind::PI ){
         pi_factor = t[0][0];
         pi = t[0][1];
-      }
+      }else if( t[0].getKind()==kind::PLUS ){
+        for( unsigned j=0; j<t[0].getNumChildren(); j++ ){
```
Comment (timothy-king) [inline src/theory/arith/arith_rewriter.cpp]: Let's split the logic here apart. Have a function that finds the index j and then logic to get pi, pi_factor and rem from j.

File: src/theory/arith/arith_rewriter.cpp [hunk 3]
```diff
@@ -409,20 +434,33 @@ RewriteResponse ArithRewriter::postRewriteTranscendental(TNode t) {
...
-          Integer two = Integer(2);
-          Integer six = Integer(6);
-          if( ra.getDenominator()==two ){
-            return RewriteResponse(REWRITE_DONE, NodeManager::currentNM()->mkConst( Rational( r.sgn() ) ) );
-          }else if( ra.getDenominator()==six ){
-            Integer five = Integer(5);
-            if( ra.getNumerator()==one || ra.getNumerator()==five ){
-              return RewriteResponse(REWRITE_DONE, NodeManager::currentNM()->mkConst( Rational( r.sgn() )/Rational(2) ) );
+          // other rational cases based on Niven's theorem (https://en.wikipedia.org/wiki/Niven%27s_theorem)
+          if( rem.isNull() ){
```
Comment (timothy-king) [inline src/theory/arith/arith_rewriter.cpp]: combine condition with else

File: src/theory/arith/arith_rewriter.cpp [hunk 4]
```diff
@@ -384,15 +384,40 @@ RewriteResponse ArithRewriter::postRewriteTranscendental(TNode t) {
...
+                remc.push_back( t[0][k] );
+              }
+            }
+            Assert( !remc.empty() );
+            rem = remc.size()==1 ? remc[0] : NodeManager::currentNM()->mkNode( kind::PLUS, remc );
+            break;
+          }
+        }
+      }      
       if( !pi_factor.isNull() ){
```
Comment (timothy-king) [inline src/theory/arith/arith_rewriter.cpp]: Move this branch into its own function.

File: src/theory/arith/arith_rewriter.cpp [hunk 5]
```diff
@@ -384,15 +384,40 @@ RewriteResponse ArithRewriter::postRewriteTranscendental(TNode t) {
...
         pi = t[0];
       }else if( t[0].getKind()==kind::MULT && t[0][0].isConst() && t[0][1].getKind()==kind::PI ){
         pi_factor = t[0][0];
         pi = t[0][1];
-      }
+      }else if( t[0].getKind()==kind::PLUS ){
+        for( unsigned j=0; j<t[0].getNumChildren(); j++ ){
+          Node monomial = t[0][j];
+          if( monomial.getKind()==kind::PI ){
+            pi_factor = NodeManager::currentNM()->mkConst(Rational(1));
```
Comment (timothy-king) [inline src/theory/arith/arith_rewriter.cpp]: Use mkRationalNode() here and elsewhere.
https://github.com/CVC4/CVC4/blob/master/src/theory/arith/arith_utilities.h#L47

File: src/theory/arith/arith_rewriter.cpp [hunk 6]
```diff
@@ -384,15 +384,40 @@ RewriteResponse ArithRewriter::postRewriteTranscendental(TNode t) {
...
+            pi = monomial;
+          }else if( monomial.getKind()==kind::MULT && monomial[0].isConst() && monomial[1].getKind()==kind::PI ){
```
Comment (timothy-king) [inline src/theory/arith/arith_rewriter.cpp]: Check that the size of the mult is 2. (Not strictly needed, but future proof.)

File: src/theory/arith/arith_rewriter.cpp [hunk 7]
```diff
@@ -384,15 +384,40 @@ RewriteResponse ArithRewriter::postRewriteTranscendental(TNode t) {
...
+            }
+            Assert( !remc.empty() );
+            rem = remc.size()==1 ? remc[0] : NodeManager::currentNM()->mkNode( kind::PLUS, remc );
+            break;
+          }
+        }
+      }      
       if( !pi_factor.isNull() ){
         Trace("arith-tf-rewrite-debug") << "Process pi factor = " << pi_factor << std::endl;
         Rational r = pi_factor.getConst<Rational>();
```
Comment (timothy-king) [inline src/theory/arith/arith_rewriter.cpp]: rename ra -> r_abs

File: src/theory/arith/arith_rewriter.cpp [hunk 8]
```diff
@@ -409,20 +434,33 @@ RewriteResponse ArithRewriter::postRewriteTranscendental(TNode t) {
...
-            Integer five = Integer(5);
-            if( ra.getNumerator()==one || ra.getNumerator()==five ){
-              return RewriteResponse(REWRITE_DONE, NodeManager::currentNM()->mkConst( Rational( r.sgn() )/Rational(2) ) );
+          // other rational cases based on Niven's theorem (https://en.wikipedia.org/wiki/Niven%27s_theorem)
+          if( rem.isNull() ){
+            Integer one = Integer(1);
+            Integer two = Integer(2);
+            Integer six = Integer(6);
+            if( ra.getDenominator()==two ){
+              Assert( ra.getNumerator()==one );
```
Comment (timothy-king) [inline src/theory/arith/arith_rewriter.cpp]: Why does this assertion hold?

File: src/theory/arith/arith_rewriter.cpp (hunk 2 above)
Comment (timothy-king) [reply src/theory/arith/arith_rewriter.cpp]: Actually even better add a struct for the triple pi_factor, pi, and rem and have a function that returns values for the triple.

**Comment by ajreynol** [comment]
For most of the comments:

I agree it would be nice to have more utilities for doing typical things in arithmetic. For quantifiers, I use a set of utility functions for arithmetic here:
https://github.com/CVC4/CVC4/blob/master/src/theory/quantifiers/quant_util.h#L87

However, I have not given them a lot of thought or optimized them.  Ideally, I would like to have a common way of doing things like for this commit (isolating a particular monomial from a sum).  

Perhaps we should open an issue to settle on a common set of arithmetic utility functions? (maybe you have some already?). Although I do recall at CAV 2015 you told me that there wasn't anything useful that is publicly available in arithmetic for doing this kind of stuff.  If needed, I can document the interface for the functions in QuantArith so that you have an idea of what is necessary for quantifiers+arithmetic.

Let me know what you think.

File: src/theory/arith/arith_rewriter.cpp (hunk 8 above)
Comment (ajreynol) [reply src/theory/arith/arith_rewriter.cpp]: ( x!=0 && abs(x/2) <1 ) => abs(x) = 1

**Comment by ajreynol** [comment]
This is part 1/3 for merging https://github.com/ajreynol/CVC4/tree/nltTfTangentPlanes-v2

File: src/theory/arith/arith_rewriter.cpp [hunk 9]
```diff
@@ -384,45 +385,105 @@ RewriteResponse ArithRewriter::postRewriteTranscendental(TNode t) {
...
-        }else if( ra == rone ){
-          return RewriteResponse(REWRITE_DONE, NodeManager::currentNM()->mkConst(Rational(0)));
-        }else{
+          Node new_arg =
+              NodeManager::currentNM()->mkNode(kind::MULT, new_pi_factor, pi);
+          if (!rem.isNull())
+          {
+            new_arg =
+                NodeManager::currentNM()->mkNode(kind::PLUS, new_arg, rem);
+          }
```
Comment (HanielB) [inline src/theory/arith/arith_rewriter.cpp]: Shouldn't you have the creation of ``new_arg`` on an ``if ... else ...``? If ``!rem.isNull()`` holds you have a spurious node construction above.

File: src/theory/arith/arith_rewriter.cpp [hunk 10]
```diff
@@ -384,45 +385,105 @@ RewriteResponse ArithRewriter::postRewriteTranscendental(TNode t) {
...
+          // sin( 2*n*PI + x ) = sin( x )
+          return RewriteResponse(
+              REWRITE_AGAIN_FULL,
+              NodeManager::currentNM()->mkNode(kind::SINE, new_arg));
+        }
+        else if (r_abs == rone)
```
Comment (HanielB) [inline src/theory/arith/arith_rewriter.cpp]: No-else-after-return

File: src/theory/arith/arith_rewriter.cpp [hunk 11]
```diff
@@ -384,45 +385,105 @@ RewriteResponse ArithRewriter::postRewriteTranscendental(TNode t) {
...
+          else
+          {
+            return RewriteResponse(
+                REWRITE_AGAIN_FULL,
+                NodeManager::currentNM()->mkNode(
+                    kind::UMINUS,
+                    NodeManager::currentNM()->mkNode(kind::SINE, rem)));
+          }
+        }
+        else if (rem.isNull())
```
Comment (HanielB) [inline src/theory/arith/arith_rewriter.cpp]: ditto

File: src/theory/arith/arith_rewriter.cpp [hunk 12]
```diff
@@ -384,45 +385,105 @@ RewriteResponse ArithRewriter::postRewriteTranscendental(TNode t) {
...
-          if( ra.getDenominator()==two ){
-            return RewriteResponse(REWRITE_DONE, NodeManager::currentNM()->mkConst( Rational( r.sgn() ) ) );
-          }else if( ra.getDenominator()==six ){
+          if (r_abs.getDenominator() == two)
+          {
+            Assert(r_abs.getNumerator() == one);
+            return RewriteResponse(REWRITE_DONE,
+                                   mkRationalNode(Rational(r.sgn())));
+          }
+          else if (r_abs.getDenominator() == six)
```
Comment (HanielB) [inline src/theory/arith/arith_rewriter.cpp]: ditto

File: src/theory/arith/arith_rewriter.cpp (hunk 9 above)
Comment (ajreynol) [reply src/theory/arith/arith_rewriter.cpp]: new_arg is being replaced by itself + rem here, so the Node construction is not spurious.

**Pull request merged**